generateGraph.ts

```diff
@@ -1,157 +1,456 @@
+                controller.down.id,
+                () => {
+                    this.yScrollOffset = Math.max(this.yScrollOffset - GRAPH_Y_AXIS_SCROLL_RATE, -(this.windowHeight + 40))
+                    if (this.yScrollOffset <= -60) {
+                        this.uiState = UI_STATE.SENSOR_SELECTION
+                        this.currentlySelectedSensorIndex = Math.abs(this.yScrollOffset + 60) / GRAPH_Y_AXIS_SCROLL_RATE
+                    }
+                    else {
+                        this.uiState = UI_STATE.GRAPH
+                    }
+                    this.processReadings();
+                    this.update() // For fast response to the above change
+                }
+            )
+
+            control.onEvent(
+                ControllerButtonEvent.Pressed,
+                controller.left.id,
+                () => {
+                    this.xScrollOffset = Math.max(0, this.xScrollOffset - 1)
+                    this.getNextDataChunk()
+                    this.processReadings()
+                    this.update() // For fast response to the above changes
+                }
+            )
+
+            control.onEvent(
+                ControllerButtonEvent.Pressed,
+                controller.right.id,
+                () => {
+                    basic.showNumber(6)
+                    const dataRows = this.dataRows
+                    this.xScrollOffset += 1
+                    this.getNextDataChunk()
+                    basic.showNumber(7)
+
+                    // basic.showNumber(x)
+                    if (this.dataRows.length > 1) {
+                        this.xScrollOffset += 1
+                        this.processReadings();
+                        this.update(); // For fast response to the above changes
+                    }
+
+                    else {
+                        this.xScrollOffset -= 1
+                        this.dataRows = dataRows
+                    }
+                }
+            )
+
+            // Select/Deselect a sensor to be drawn:
+            control.onEvent(
+                ControllerButtonEvent.Pressed,
+                controller.A.id,
+                () => {
+                    if (this.uiState == UI_STATE.SENSOR_SELECTION) {
+                        this.drawSensorStates[this.currentlySelectedSensorIndex] = !this.drawSensorStates[this.currentlySelectedSensorIndex]
+                    }
+                }
+            )
+        }
+
+        /**
+         * Used to retrieve the next chunk of data
+         * Can be used to intialise this.dataRows
+         * Invoked when this.xScrollOffset changes (Left or Right is pressed)
+         *@returns dataRows where each row contains 4 columns [Sensor Name, Time(ms), Reading, Events]
+         */
+        private getNextDataChunk() {
+            this.dataRows = []
+            const tokens = datalogger.getRows(this.xScrollOffset * 10, (this.xScrollOffset * 10) + 10).split("_");

```

Comment:
A general guideline: Code should be as self-explanatory as possible, so it can be understood at a glance.
```suggestion
            const startIndex = this.xScrollOffset * 10;
            const rows = datalogger.getRows(startIndex, startIndex + 10);
            const tokens = rows.split("_");

```